File: SmartBusTracker/src/utils/kalmanFilter.ts
/**
 * A simple 1D Kalman filter implementation to smooth out noisy GPS coordinates.
 */
export class KalmanFilter {
    private R: number; // measurement noise variance
    private Q: number; // process noise variance
    private P: number; // estimation error covariance
    private X: number; // value estimate
    private initialized: boolean;

    constructor(R = 1, Q = 0.1) {
        this.R = R;
        this.Q = Q;
        this.P = 1;
        this.X = 0;
        this.initialized = false;
    }

    /**
     * Filters a new measurement.
     * @param measurement The noisy coordinate measurement
     * @returns The smoothed coordinate estimate
     */
    filter(measurement: number): number {
        if (!this.initialized) {
            this.X = measurement;
            this.initialized = true;
            return this.X;
        }

        // Prediction update
        this.P = this.P + this.Q;

        // Measurement update
        const K = this.P / (this.P + this.R); // Kalman gain
        this.X = this.X + K * (measurement - this.X);
        this.P = (1 - K) * this.P;

        return this.X;
    }

    reset(): void {
        this.P = 1;
        this.X = 0;
        this.initialized = false;
    }
}

File: smartbus-backend/app/services/kalman_filter.py
import numpy as np

class KalmanFilterGPS:
    """
    2D Kalman Filter for smoothing noisy GPS coordinates.
    State vector: [lat, lng, vel_lat, vel_lng]
    """
    def __init__(self):
        self.dt = 5.0        # time step in seconds
        self.initialized = False
        
        # State vector [lat, lng, vel_lat, vel_lng]
        self.x = np.zeros((4, 1))
        
        # State transition matrix
        self.F = np.array([
            [1, 0, self.dt, 0],
            [0, 1, 0, self.dt],
            [0, 0, 1, 0],
            [0, 0, 0, 1]
        ])
        
        # Measurement matrix (we only measure lat, lng)
        self.H = np.array([
            [1, 0, 0, 0],
            [0, 1, 0, 0]
        ])
        
        # Measurement noise covariance (GPS accuracy ~5m)
        self.R = np.eye(2) * 0.0001
        
        # Process noise covariance
        self.Q = np.eye(4) * 0.00001
        
        # Estimation error covariance
        self.P = np.eye(4) * 1.0
    
    def filter(self, lat: float, lng: float) -> tuple[float, float]:
        """
        Apply Kalman filter to GPS measurement.
        Returns smoothed (lat, lng) tuple.
        """
        measurement = np.array([[lat], [lng]])
        
        if not self.initialized:
            self.x[0] = lat
            self.x[1] = lng
            self.initialized = True
            return lat, lng
        
        # Prediction step
        self.x = self.F @ self.x
        self.P = self.F @ self.P @ self.F.T + self.Q
        
        # Update step
        y = measurement - self.H @ self.x
        S = self.H @ self.P @ self.H.T + self.R
        K = self.P @ self.H.T @ np.linalg.inv(S)
        self.x = self.x + K @ y
        self.P = (np.eye(4) - K @ self.H) @ self.P
        
        return float(self.x[0]), float(self.x[1])
    
    def reset(self):
        self.initialized = False
        self.x = np.zeros((4, 1))
        self.P = np.eye(4) * 1.0

# Store one filter instance per bus
kalman_filters: dict[str, KalmanFilterGPS] = {}

def get_kalman_filter(bus_id: str) -> KalmanFilterGPS:
    if bus_id not in kalman_filters:
        kalman_filters[bus_id] = KalmanFilterGPS()
    return kalman_filters[bus_id]
